lib/statement_identifier_service.rb

```diff
@@ -0,0 +1,167 @@
+# frozen_string_literal: true
+
+require 'mpi/service'
+
+class StatementIdentifierService
+  include SentryLogging
+  STATSD_KEY_PREFIX = 'api.copay_notifications.new_statement'
+
+  class MalformedMCPStatement < StandardError; end
+  class UnableToSourceEmailForStatement < StandardError; end
+
+  class ProfileMissingEmail < StandardError
+    def initialize(vet360_id)
+      @vet360_id = vet360_id
+      message = "ProfileMissingEmail: Unable to derive an email address from vet360 id: #{@vet360_id}"
+      super(message)
+    end
+  end
+
+  class Vet360IdNotFound < StandardError
+    def initialize(icn)
+      @icn = icn
+      message = "MPIProfileMissingVet360Id: MPI Profile is missing vet360id #{@icn}"
+      super(message)
+    end
+  end
+
+  def initialize(statement)
+    @statement = statement
+    raise MalformedMCPStatement, statement unless legal_statement
+
+    @identifier = @statement['veteranIdentifier']
+    @identifier_type = @statement['identifierType']
+    @facility_id = @statement['facilityNum']
+    @identifier_key = nil
+  end
+
+  def derive_email_address
+    email_address = if @identifier_type == 'edipi'
+                      source_from_edipi
+                    else
+                      source_from_vista
+                    end
+    raise UnableToSourceEmailForStatement, @statement unless email_address
+
+    email_address
+  end
+
+  def vista_account_id
+    offset = 16 - (@facility_id + @identifier).length
+    padding = '0' * offset if offset >= 0
+    "#{@facility_id}#{padding}#{@identifier}"
+  end
+
+  private
+
+  def source_from_edipi
+    @identifier_key = edipi_key
+    index = IdentifierIndex.find(@identifier_key)
+    return index.email_address if index
+
+    credential_email = edipi_email_address(@identifier)
+    if credential_email
+      save_to_identifier(@identifier_key, credential_email)
+      return credential_email
+    end
+    source_from_mpi
+  end
+
+  def source_from_vista
+    @identifier_key = vista_key
+    index = IdentifierIndex.find(vista_key)
+    return index.email_address if index
+
+    source_from_mpi
+  end
+
+  def source_from_mpi
+    mpi_response = get_mpi_profile
+    if mpi_response.ok?
+      StatsD.increment("#{STATSD_KEY_PREFIX}.mpi.success")
+      icn_email = icn_email_address(mpi_response.profile.icn)
+      if icn_email
+        save_to_identifier(@identifier_key, icn_email)
+        return icn_email
+      end
+      vet360_id = mpi_response&.profile&.vet360_id
```

Comment:
I still think we might want to use something like `VANotify::IcnJob` here with the ICN from MPI to offload the profile email lookup to VA Notify's service but that can be discovered for or implemented in a separate ticket.

Reply:
Very happy to implement in a future ticket 👌